File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/render/NativePointer.java
package org.maplibre.nativeffi.render;

import java.util.Objects;

/**
 * Opaque borrowed native address value used for backend interop handles.
 *
 * <p>The pointer transfers no ownership and grants no general memory access. Callers that create a
 * pointer from an address keep the backend object alive and externally synchronized for the full C
 * API borrow window of the render descriptor or active frame that receives it.
 */
public final class NativePointer {
  public static final NativePointer NULL = new NativePointer(0, null);

  private final long address;
  private final FrameScope scope;

  private NativePointer(long address, FrameScope scope) {
    this.address = address;
    this.scope = scope;
  }

  public static NativePointer ofAddress(long address) {
    return address == 0 ? NULL : new NativePointer(address, null);
  }

  static NativePointer scoped(long address, FrameScope scope) {
    return address == 0 ? NULL : new NativePointer(address, Objects.requireNonNull(scope, "scope"));
  }

  public long address() {
    ensureActive();
    return address;
  }

  public boolean isNull() {
    ensureActive();
    return address == 0;
  }

  private void ensureActive() {
    if (scope != null) {
      scope.ensureActive();
    }
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof NativePointer that && address() == that.address();
  }

  @Override
  public int hashCode() {
    return Long.hashCode(address());
  }

  @Override
  public String toString() {
    return "NativePointer[address=0x" + Long.toHexString(address()) + "]";
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/render/NativePointer.kt
package org.maplibre.nativeffi.render

/** Opaque borrowed native address value used for backend interop handles. */
public class NativePointer
private constructor(private val addressValue: Long, private val scope: FrameScope?) {
  /** Borrowed address bit pattern. Access rejects use after a scoped frame closes. */
  public val address: Long
    get() {
      ensureActive()
      return addressValue
    }

  /** Returns true when this pointer represents a null backend handle. */
  public val isNull: Boolean
    get() {
      ensureActive()
      return addressValue == 0L
    }

  override fun equals(other: Any?): Boolean {
    ensureActive()
    if (other !is NativePointer) return false
    other.ensureActive()
    return addressValue == other.addressValue
  }

  override fun hashCode(): Int {
    ensureActive()
    return addressValue.hashCode()
  }

  override fun toString(): String {
    ensureActive()
    return "NativePointer[address=0x${addressValue.toString(16)}]"
  }

  private fun ensureActive() {
    scope?.ensureActive()
  }

  public companion object {
    /** Null native pointer value. */
    public val NULL: NativePointer = NativePointer(0L, null)

    /**
     * Creates an opaque borrowed backend interop pointer from an address bit pattern.
     *
     * The caller retains ownership of the pointed-to backend resource and keeps it valid and
     * synchronized for the full C API borrow window documented by the descriptor that receives this
     * value. This value grants no general memory access.
     */
    public fun ofAddress(address: Long): NativePointer =
      if (address == 0L) NULL else NativePointer(address, null)

    internal fun scoped(address: Long, scope: FrameScope): NativePointer =
      if (address == 0L) NULL else NativePointer(address, scope)
  }
}
